### PR TITLE
DBZ-3557 Add `snapshot.locking.mode` option for Oracle

### PR DESCRIPTION
**Pull request by Naros**
https://issues.redhat.com/browse/DBZ-3557

### CODE REVIEW

File: debezium-connector-oracle/src/main/java/io/debezium/connector/oracle/OracleConnectorConfig.java [hunk 1]
```diff
@@ -95,6 +91,19 @@
...
+    public static final Field SNAPSHOT_LOCKING_MODE = Field.create("snapshot.locking.mode")
+            .withDisplayName("Snapshot locking mode")
+            .withEnum(SnapshotLockingMode.class, SnapshotLockingMode.EXCLUSIVE)
+            .withWidth(Width.SHORT)
+            .withImportance(Importance.LOW)
+            .withDescription("Controls how the connector holds locks on tables while performing the schema snapshot. The default is 'exclusive', "
+                    + "which means the connector will hold an exclusive table lock (prevents any updates) for just the initial portion of the snapshot "
+                    + "while the database schemas and other metadata are being read. The remaining work in a snapshot involves selecting all rows from "
+                    + "each table, and this is done using a flashback query that requires no locks. However, in some cases it may be desirable to allow "
+                    + "concurrent access to the table but prevent locking the entire table; in such ases set this property to 'shared'. Using a value of "
```
Comment (ant0nk) [inline debezium-connector-oracle/src/main/java/io/debezium/connector/oracle/OracleConnectorConfig.java:103]: Misprint - "ases" instead of "cases"

Comment (Naros) [reply debezium-connector-oracle/src/main/java/io/debezium/connector/oracle/OracleConnectorConfig.java:103]: Thanks @ant0nk!

File: debezium-connector-oracle/src/main/java/io/debezium/connector/oracle/OracleSnapshotChangeEventSource.java [hunk 2]
```diff
@@ -86,24 +86,32 @@ protected SnapshotContext prepare(ChangeEventSourceContext context) throws Excep
...
+                    if (!sourceContext.isRunning()) {
+                        throw new InterruptedException("Interrupted while locking table " + tableId);
+                    }
+
+                    LOGGER.debug("Locking table {}", tableId);
+                    if (connectorConfig.getSnapshotLockingMode().usesExclusiveLocking()) {
+                        statement.execute("LOCK TABLE " + quote(tableId) + " IN EXCLUSIVE MODE");
+                    }
+                    else {
+                        statement.execute("LOCK TABLE " + quote(tableId) + " IN SHARE UPDATE MODE");
```
[the remark this answers is not in the document]
Comment (Naros) [reply debezium-connector-oracle/src/main/java/io/debezium/connector/oracle/OracleSnapshotChangeEventSource.java:103]: Yes a DDL change would attempt to acquire an exclusive lock and would be prevented.  

The difference as I understand it is that `EXCLUSIVE` prevents any user from doing any operation on the table, including insert, update, delete, select for update, as well as other lock table statements where-as the share update / row share mode is less restrictive and will allow other operations to occur such as inserts, updates, deletes, and select for updates that are permissible with a shared lock.  

As for `SHARE UPDATE`, I'll change that to `ROW SHARE` to use the newer syntax.

[the remark this answers is not in the document]
Comment (Naros) [reply debezium-connector-oracle/src/main/java/io/debezium/connector/oracle/OracleSnapshotChangeEventSource.java:103]: > If so, is there a need for EXCLUSIVE at all?

Coming back to this, I suppose in hindsight there really isn't.  `ROW SHARE` should suffice and relaxes the lock restrictions, so perhaps that's the more simplier fix all together.  Would you agree?

[the remark this answers is not in the document]
Comment (gunnarmorling) [reply debezium-connector-oracle/src/main/java/io/debezium/connector/oracle/OracleSnapshotChangeEventSource.java:103]: Ok, so if you're positive that using `ROW SHARE` will prevent a concurrent table schema change from happening, let's use this, and not even expose `EXCLUSIVE` in that new snapshot mode enum to begin with. I.e. we'd only have the two values `ROW_SHARE` and `NONE`.

**Comment by Naros** [comment]
Looks like a small build failure, I'll fix that shortly.  Didn't expect that method to be used in tests :)

File: documentation/modules/ROOT/pages/connectors/oracle.adoc [hunk 3]
```diff
@@ -30,7 +30,7 @@ the default behavior for performing a snapshot consists of the following steps.
 you can change this behavior by setting the `snapshot.mode` connector configuration property to a value other than `initial`.
 
 1. Determine the tables to be captured
-2. Obtain an `IN EXCLUSIVE MODE` lock on each of the monitored tables to ensure that no structural changes can occur to any of the tables.
+2. Obtain an `ROW SHARE MODE` lock on each of the monitored tables to ensure that no structural changes can occur to any of the tables.
```
Comment (gunnarmorling) [inline documentation/modules/ROOT/pages/connectors/oracle.adoc:33]:
```suggestion
2. Obtain a `ROW SHARE MODE` lock on each of the monitored tables to ensure that no structural changes can occur to any of the tables.
```

File: documentation/modules/ROOT/pages/connectors/oracle.adoc [hunk 4]
```diff
@@ -1301,6 +1301,14 @@ Skipping should be used only with care as it can lead to data loss or mangling w
 |_initial_
 |A mode for taking an initial snapshot of the structure and optionally data of captured tables. Supported values are _initial_ (will take a snapshot of structure and data of captured tables; useful if topics should be populated with a complete representation of the data from the captured tables) and _schema_only_ (will take a snapshot of the structure of captured tables only; useful if only changes happening from now onwards should be propagated to topics). Once the snapshot is complete, the connector will continue reading change events from the database's redo logs.
 
+|[[oracle-property-snapshot-locking-mode]]<<oracle-property-snapshot-locking-mode, `+snapshot.locking.mode+`>>
+|_shared_
+a|Controls whether and how long the connector holds a table lock, which prevents certain types of changes table operations, while the connector is performing a snapshot. Possible settings are: +
+ +
+`shared` - allows concurrent access to the table but prevents any session from acquiring a table exclusive lock. +
```
Comment (gunnarmorling) [inline documentation/modules/ROOT/pages/connectors/oracle.adoc:1308]:
```suggestion
`shared` - allows concurrent access to the table but prevents any session from acquiring a table exclusive lock (specifically, the connector will acquire a `ROW SHARE`level lock while capturing the schemas of the tables). +
```

**Review by gunnarmorling — APPROVED** [review]
LGTM; will apply the doc change and merge. Thx!

**Pull request merged**